File: MemberApp/Models/Channel.swift
import SwiftUI

/// An announcement channel a member can follow.
///
/// The mock modelled these as push subscriptions. There is no push
/// infrastructure and no channel backend yet, so for now a channel is a
/// *display* preference — which announcements a member wants to see. The same
/// stored set will drive notifications when there is something to send them.
/// Not Hashable: `LocalizedStringKey` is not, and identity here is `id` anyway.
struct Channel: Identifiable {
    /// Stable key for persistence. Never localised — it is stored, not shown.
    let id: String
    /// SF Symbol, where one carries the meaning. Nil for schools, whose
    /// abbreviation *is* the recognisable mark.
    let symbol: String?
    /// Fallback badge text when `symbol` is nil.
    let badge: String
    /// Distinguishes the rows at a glance. School colours approximate each
    /// institution's palette — only NTU's is taken from its own site
    /// (theme-color #D71440); the others are close, not official.
    let tint: Color
    let name: LocalizedStringKey
    let detail: LocalizedStringKey
    /// School channels are matched against `Profile.school` to preselect.
    let school: String?
}

extension Channel {
    static let all: [Channel] = [
        .init(id: "all", symbol: "megaphone.fill", badge: "STSA",
              tint: Theme.Palette.brand,
              name: "全體公告", detail: "所有會員都會收到的公告",
              school: nil),
        .init(id: "freshmen", symbol: "graduationcap.fill", badge: "新生",
              tint: Color(red: 0.05, green: 0.52, blue: 0.49),
              name: "新生資訊", detail: "迎新、接機、住宿與開學前準備",
              school: nil),
        .init(id: "nus", symbol: nil, badge: "NUS",
              tint: Color(red: 0.94, green: 0.49, blue: 0.00),
              name: "NUS 資訊", detail: "National University of Singapore",
              school: "NUS"),
        .init(id: "ntu", symbol: nil, badge: "NTU",
              tint: Color(red: 0.84, green: 0.08, blue: 0.25),
              name: "NTU 資訊", detail: "Nanyang Technological University",
              school: "NTU"),
        .init(id: "smu", symbol: nil, badge: "SMU",
              tint: Color(red: 0.02, green: 0.22, blue: 0.42),
              name: "SMU 資訊", detail: "Singapore Management University",
              school: "SMU"),
    ]

    /// Sensible starting point: everyone gets the all-members channel and the
    /// freshman feed, plus their own school if the profile knows it.
    static func defaultSubscriptions(school: String?) -> Set<String> {
        var ids: Set<String> = ["all", "freshmen"]
        if let school, let match = all.first(where: { $0.school == school }) {
            ids.insert(match.id)
        }
        return ids
    }
}

File: android/app/src/main/java/tw/stsa/memberapp/model/Channel.kt
package tw.stsa.memberapp.model

import androidx.annotation.StringRes
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Campaign
import androidx.compose.material.icons.filled.School
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import tw.stsa.memberapp.R
import tw.stsa.memberapp.designsystem.Theme

/**
 * An announcement channel a member can follow.
 *
 * The mock modelled these as push subscriptions. There is no push
 * infrastructure and no channel backend yet, so for now a channel is a
 * *display* preference — which announcements a member wants to see. The same
 * stored set will drive notifications when there is something to send them.
 */
data class Channel(
    /** Stable key for persistence. Never localised — it is stored, not shown. */
    val id: String,
    /**
     * Icon, where one carries the meaning. Null for schools, whose abbreviation
     * *is* the recognisable mark.
     */
    val icon: ImageVector?,
    /** Fallback badge text when [icon] is null. A proper noun, so not localised. */
    val badge: String,
    /**
     * Distinguishes the rows at a glance. School colours approximate each
     * institution's palette — only NTU's is taken from its own site
     * (theme-color #D71440); the others are close, not official.
     */
    val tint: Color,
    @StringRes val nameRes: Int,
    @StringRes val detailRes: Int,
    /** School channels are matched against `Profile.school` to preselect. */
    val school: String?,
) {
    companion object {
        val all: List<Channel> = listOf(
            Channel(
                id = "all",
                icon = Icons.Filled.Campaign,
                badge = "STSA",
                tint = Theme.Brand,
                nameRes = R.string.channel_all,
                detailRes = R.string.channel_all_detail,
                school = null,
            ),
            Channel(
                id = "freshmen",
                icon = Icons.Filled.School,
                badge = "新生",
                tint = Color(0xFF0D857D),
                nameRes = R.string.channel_freshmen,
                detailRes = R.string.channel_freshmen_detail,
                school = null,
            ),
            Channel(
                id = "nus",
                icon = null,
                badge = "NUS",
                tint = Color(0xFFF07D00),
                nameRes = R.string.channel_nus,
                detailRes = R.string.channel_nus_detail,
                school = "NUS",
            ),
            Channel(
                id = "ntu",
                icon = null,
                badge = "NTU",
                tint = Color(0xFFD71440),
                nameRes = R.string.channel_ntu,
                detailRes = R.string.channel_ntu_detail,
                school = "NTU",
            ),
            Channel(
                id = "smu",
                icon = null,
                badge = "SMU",
                tint = Color(0xFF05386B),
                nameRes = R.string.channel_smu,
                detailRes = R.string.channel_smu_detail,
                school = "SMU",
            ),
        )

        /**
         * Sensible starting point: everyone gets the all-members channel and the
         * freshman feed, plus their own school if the profile knows it.
         */
        fun defaultSubscriptions(school: String?): Set<String> {
            val ids = mutableSetOf("all", "freshmen")
            val match = school?.let { name -> all.firstOrNull { it.school == name } }
            if (match != null) ids.add(match.id)
            return ids
        }
    }
}
